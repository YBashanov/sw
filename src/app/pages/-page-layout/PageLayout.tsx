import React from 'react';
import { PREFIX_CLS } from '@/constants';
import Separator from '@/components/Separator';
import ThemeControls from '@/themes/ThemeControls';
import LanguageControls from '@/i18n/LanguageControls';

import './index.less';

const mainClassName = `${PREFIX_CLS}-page-layout`;

const PageLayout = (Page: React.ElementType) => {
    return function WrappedPage(): React.ReactElement {
        return (
            <div className={`${mainClassName}`}>
                <div className={`${mainClassName}-content`}>
                    <div className={`${mainClassName}-content-page`}>
                        <Page />
                    </div>
                </div>
                <div className={`${mainClassName}-footer`}>
                    <Separator />
                    <ThemeControls />
                    <LanguageControls />
                </div>
            </div>
        );
    };
};

export default PageLayout;
