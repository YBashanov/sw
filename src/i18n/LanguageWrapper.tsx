import React, { createContext, useState } from 'react';
import { getIntlLocale, getIntlMessages } from '@/i18n';
import { IntlProvider } from 'react-intl';

export type Language = 'ru' | 'en';

export const defaultLanguage: Language = 'ru';

type LanguageContextType = {
    language?: Language;
    setLanguage?: (language: Language) => void;
};

type LanguageWrapperType = {
    children: React.ReactNode;
};

export const LanguageContext = createContext<LanguageContextType>({
    language: defaultLanguage,
});

export const defaultSaveLang = localStorage.getItem('language') as Language;

const LanguageWrapper: React.FC<LanguageWrapperType> = ({ children }) => {
    const [language, setLanguage] = useState<Language>(defaultSaveLang || defaultLanguage);

    const handleSetLanguage = (lang: Language) => {
        localStorage.setItem('language', lang);

        // установка для React-компонентов
        setLanguage(lang);

        // установка для статики (схемы Yup, модели Redux)
        // устанавливается в i18n/index

        // требуется перезагрузка для сохранения в localStorage и взятия оттуда
        location.href = '/';
    };

    // тестирование смены языка. (в идеале это должно быть по кнопке)
    React.useEffect(() => {
        // const wantLang = 'ru';
        // if (defaultSaveLang !== wantLang) handleSetLanguage(wantLang);
    }, []);

    return (
        <LanguageContext.Provider value={{ language, setLanguage: handleSetLanguage }}>
            <IntlProvider locale={getIntlLocale(language)} messages={getIntlMessages(language)}>
                {children}
            </IntlProvider>
        </LanguageContext.Provider>
    );
};

export default LanguageWrapper;
