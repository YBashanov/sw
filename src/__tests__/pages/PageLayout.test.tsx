import React from 'react';
import { describe, expect, it } from '@jest/globals';
import { render } from 'enzyme';
import PageLayout from '@/app/pages/-page-layout/PageLayout';
import LanguageWrapper from '@/i18n/LanguageWrapper';

const TestPage = () => {
    return <div id="unique_key">TestPage</div>;
};

describe('testing: PageLayout', function () {
    const Wrapper = PageLayout(TestPage);
    const JSX = (
        <LanguageWrapper>
            <Wrapper />
        </LanguageWrapper>
    );
    const Component = render(JSX);

    it('Snapshot correctly', function () {
        expect(Component).toMatchSnapshot();
    });

    it('Static HTML', function () {
        expect(Component.text()).toEqual('TestPageДополнительноВыбор темыСтандартРозовыйЗеленыйВыбор языкаRUEN');
    });

    it('Checking the added content', function () {
        expect(Component.find('#unique_key')).toHaveLength(1);
    });
});
