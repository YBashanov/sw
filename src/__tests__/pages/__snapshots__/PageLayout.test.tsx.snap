// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing: PageLayout Snapshot correctly 1`] = `
LoadedCheerio {
  "0": <div
    class="sw-page-layout"
  >
    <div
      class="sw-page-layout-content"
    >
      <div
        class="sw-page-layout-content-page"
      >
        <div
          id="unique_key"
        >
          TestPage
        </div>
      </div>
    </div>
    <div
      class="sw-page-layout-footer"
    >
      <div
        class="separator"
      >
        <div
          class="label"
        >
          Дополнительно
        </div>
      </div>
      <div
        class="theme-controls"
      >
        <div
          class="label"
        >
          Выбор темы
        </div>
        <div
          class="controls"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-primary"
            type="button"
          >
            <span>
              Стандарт
            </span>
          </button>
          <button
            class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-default"
            type="button"
          >
            <span>
              Розовый
            </span>
          </button>
          <button
            class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-default"
            type="button"
          >
            <span>
              Зеленый
            </span>
          </button>
        </div>
      </div>
      <div
        class="language-controls"
      >
        <div
          class="label"
        >
          Выбор языка
        </div>
        <div
          class="controls"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-primary"
            id="ru"
            type="button"
          >
            <span>
              RU
            </span>
          </button>
          <button
            class="ant-btn css-dev-only-do-not-override-ed5zg0 ant-btn-default"
            id="en"
            type="button"
          >
            <span>
              EN
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
